xarray/core/accessor_str.py

```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+
+        See Also
+        --------
+        DataArray.str.extract
+        DataArray.str.findall
+        re.compile
+        re.findall
+        pandas.Series.str.extractall
+        """
+        pat = self._re_compile(pat, flags, case)
+
+        if pat.groups == 0:
+            raise ValueError("No capture groups found in pattern.")
+
+        if group_dim in self._obj.dims:
+            raise KeyError(f"Group dimension {group_dim} already present in DataArray.")
+
+        if match_dim in self._obj.dims:
+            raise KeyError(f"Match dimension {match_dim} already present in DataArray.")
+
+        if group_dim == match_dim:
+            raise KeyError(
+                f"Group dimension {group_dim} is the same as match dimension {match_dim}."
+            )
+
+        _get_count = lambda x: len(pat.findall(x))
+        maxcount = self._apply(_get_count, dtype=np.int_).max().data.tolist()
+
+        def _get_res(val, pat=pat, num=maxcount, dtype=self._obj.dtype):
+            matches = pat.findall(val)
+            res = np.zeros([maxcount, pat.groups], dtype)
+
+            if pat.groups == 1:
+                for imatch, match in enumerate(matches):
+                    res[imatch, 0] = match
+            else:
+                for imatch, match in enumerate(matches):
+                    for jmatch, submatch in enumerate(match):
+                        res[imatch, jmatch] = submatch
+
+            return res
+
+        return self._apply(
+            # dtype MUST be object or strings can be truncated
+            # See: https://github.com/numpy/numpy/issues/8352
+            _get_res,
+            dtype=np.object_,
+            output_core_dims=[[group_dim, match_dim]],
+            output_sizes={group_dim: pat.groups, match_dim: maxcount},
+        ).astype(self._obj.dtype.kind)
+
+    def findall(
+        self,
+        pat: Any,
+        case: bool = None,
+        flags: int = 0,
+    ) -> Any:
+        """
+        Find all occurrences of pattern or regular expression in the DataArray.
+
+        Equivalent to applying re.findall() to all the elements in the DataArray.
+        Results in an object array of lists.
+        If there is only one capture group, the lists will be a sequence of matches.
+        If there are multiple capture groups, the lists will be a sequence of lists,
+        each of which contains a sequence of matches.
+
+        Parameters
+        ----------
+        pat : str or re.Pattern
+            A string contain a regular expression or
+            a compiled regular expression object.
+        case : bool, default: True
+            If True, case sensitive.
+            Cannoy be set if `pat` is a compiled regular expression.
+        flags : int, default: 0
+            - re module flags, e.g. re.IGNORECASE. Use ``0`` for no flags.
+            - Cannot be set if `pat` is a compiled regex
+
+        Returns
+        -------
+        extracted : same type as values or object array
```

Comment:
That will have `dtype=object` right? Also the ones below.

Reply:
Fixed.  This is the only one that only returns object arrays, though